### ISSUE TITLE
CORS: OPTIONS call not handled

### ISSUE THREAD

**Issue by pierreprinetti** [gorilla/mux#381]
**What version of Go are you running?**
go version go1.10.2 darwin/amd64

**What version of gorilla/mux are you at?**
c85619274f5d55a67491010f883a91f568b3a9b0

**Describe your problem**

* Create a router
* Register a route with `.Methods("GET")`
* Make the router use the new CORSMiddleware
* Call `OPTIONS` on the route
* Expect a `200 OK` with the `Access-Control-Allowed-Methods` set to `GET,OPTIONS`; get instead a 405.

The code that is supposed to handle `OPTIONS` requests (below) is never reached, unless `OPTIONS` is a registered method for the route.

https://github.com/gorilla/mux/blob/c85619274f5d55a67491010f883a91f568b3a9b0/middleware.go#L64-L66

**Paste a minimal, runnable, reproduction of your issue below**

Add this test to `middleware_test.go`
```Go
func TestCORSMiddlewareOPTIONS(t *testing.T) {
	router := NewRouter()

	cases := []struct {
		path                   string
		response               string
		method                 string
		testURL                string
		expectedAllowedMethods string
	}{
		{"/g/{o}", "a", "POST", "/g/asdf", "POST,PUT,GET,OPTIONS"},
		{"/g/{o}", "b", "PUT", "/g/bla", "POST,PUT,GET,OPTIONS"},
		{"/g/{o}", "c", "GET", "/g/orilla", "POST,PUT,GET,OPTIONS"},
		{"/g", "d", "POST", "/g", "POST,OPTIONS"},
	}

	for _, tt := range cases {
		router.HandleFunc(tt.path, stringHandler(tt.response)).Methods(tt.method)
	}

	router.Use(CORSMethodMiddleware(router))

	for _, tt := range cases {
		rr := httptest.NewRecorder()
		req := newRequest("GET", tt.testURL)

		router.ServeHTTP(rr, req)

		allowedMethods := rr.HeaderMap.Get("Access-Control-Allow-Methods")

		if want, have := 200, rr.Result().StatusCode; have != want {
			t.Errorf("Expected status code %d, found %d", want, have)
		}

		if allowedMethods != tt.expectedAllowedMethods {
			t.Errorf("Expected Access-Control-Allow-Methods '%s', found '%s'", tt.expectedAllowedMethods, allowedMethods)
		}
	}
}
```

Note that this new test passes if `OPTIONS` is explicitly set as a method in the test-table:
```Go
		{"/g/{o}", "a", "POST", "/g/asdf", "POST,PUT,GET,OPTIONS"},
		{"/g/{o}", "b", "PUT", "/g/bla", "POST,PUT,GET,OPTIONS"},
		{"/g/{o}", "c", "GET", "/g/orilla", "POST,PUT,GET,OPTIONS"},
		{"/g", "d", "POST", "/g", "POST,OPTIONS"},
		{"/g", "", "OPTIONS", "/g", "POST,OPTIONS"},
		{"/g/{o}", "", "OPTIONS", "/g/orilla", "POST,PUT,GET,OPTIONS"},
```

If this is the desired behaviour, and the `OPTIONS` method has to be set while registering the route in order for the middleware to handle it, then it would be better to document it. But honestly, it doesn't sound desirable.

**Comment by elithrar** [comment]
@pierreprinetti - we're looking at expanding middleware to run even without a match, as per https://github.com/gorilla/mux/pull/378

If you'd like to give that PR a cursory review/feedback, I'd appreciate. Currently traveling and may not have time to address for a few weeks.

**Comment by pierreprinetti** [comment]
@elithrar Thanks for the fast response!

Just one thing:

With the current code, setting "OPTIONS" explicitly as a verb when registering a route will have the effect of allowing the `CORSMethodMiddleware` to handle it.

I think it's a totally valid solution! If this is the intended behaviour? If it is, then maybe we could add a short line in the docs about it.

**Comment by elithrar** [comment]
Let’s definitely document that against the middleware. There are cases
where you may not want to expose routes via CORS, and that seems like a
sensible compromise without enabling too much magic.

**Comment by vcGF** [comment]
<img width="403" alt="screen shot 2018-09-03 at 5 09 02 am" src="https://user-images.githubusercontent.com/10025321/44960752-8db03e80-af37-11e8-9631-1913bea0af42.png">

Somehow I get this error, even if I have the latest release code. Hmmm

**Comment by elithrar** [comment]
Can you post a reproducible example, and include what your requests are, and what errors are returned?

**Comment by vcGF** [comment]
I m using Google App Engine to host my service.
Sending my payload through POSTMAN doesn't trigger the OPTIONS method, hence it goes straight to POST and is successful.

However, utimately the API is to be trigerred from a Angular6 frontend, and Chrome (or Safari) are sending those POST request first, the POST and the JSON payload don't even reach the server as the OPTIONS call fails.

Here is the trace from the Chrome debug tools

![screen shot 2018-09-03 at 11 53 24 am](https://user-images.githubusercontent.com/10025321/44966943-b30d6e80-af70-11e8-9b54-1e8ce177c3a3.png)

![screen shot 2018-09-03 at 11 53 58 am](https://user-images.githubusercontent.com/10025321/44966947-b7d22280-af70-11e8-928a-c7668c750717.png)


Here is how I setup the router in my Go file.

<img width="388" alt="screen shot 2018-09-03 at 11 59 39 am" src="https://user-images.githubusercontent.com/10025321/44966989-f667dd00-af70-11e8-9f2f-234a5a5a0137.png">

**Comment by vcGF** [comment]
Did I miss one step in the configuration here ?

**Comment by elithrar** [comment]
Yes: you need to explicitly allow the origin by adding an `Access-Control-Allow-Origin` header, as per the error message. `localhost` is not a valid origin.

I recommend taking a look at https://github.com/rs/cors
